File: 150 Evaluate Reverse Polish Notation/Solution.py
'''
Time Complexity: O(n) As going through entire list
Space Complexity: O(n) due to stack. but each operand reduces the stack size. However, the space usage is still proportional to the length of the input, which is n.
'''

class Solution:
    def evalRPN(self, tokens: list[str]) -> int:
        stack = []
        for char in tokens:
            if char == "+":
                stack.append(stack.pop() + stack.pop())
            elif char == "-":
                b, a = stack.pop(), stack.pop()
                stack.append(a - b)
            elif char == "*":
                stack.append(stack.pop() * stack.pop())
            elif char == "/":
                b, a = stack.pop(), stack.pop()
                stack.append(int(a / b))
            else:
                stack.append(int(char))
        return stack[-1]

if __name__ == "__main__":
    solution = Solution()
    token1 = ["2","1","+","3","*"] #9
    token2 = ["4","13","5","/","+"] #6
    token3 = ["10","6","9","3","+","-11","*","/","*","17","+","5","+"] #22
    result = solution.evalRPN(token3)
    print(result)

File: 150 Evaluate Reverse Polish Notation/Solution.java
import java.util.Stack;

class Solution150 {
    public int evalRPN(String[] tokens) {
        Stack<Integer> stack = new Stack<>();

        for (String token : tokens) {
            switch (token) {
                case "+":
                    stack.push(stack.pop() + stack.pop());
                    break;
                case "-":
                    int b1 = stack.pop();
                    int a1 = stack.pop();
                    stack.push(a1 - b1);
                    break;
                case "*":
                    stack.push(stack.pop() * stack.pop());
                    break;
                case "/":
                    int b2 = stack.pop();
                    int a2 = stack.pop();
                    stack.push(a2 / b2);  // Integer division in Java
                    break;
                default:
                    stack.push(Integer.parseInt(token));
                    break;
            }
        }
        return stack.peek();
    }

    public static void main(String[] args) {
        Solution150 solution = new Solution150();
        String[] token1 = {"2", "1", "+", "3", "*"};  // 9
        String[] token2 = {"4", "13", "5", "/", "+"}; // 6
        String[] token3 = {"10", "6", "9", "3", "+", "-11", "*", "/", "*", "17", "+", "5", "+"}; // 22
        int result = solution.evalRPN(token3);
        System.out.println(result);
    }
}